### PR TITLE
Add a json example using Gson

### PR DESCRIPTION
**Pull request by windie**
Motivation:

Add a codec implementation example using Gson. See #1129.

Modifications:

Add an echo sever and client that decode/encode json using Gson.

Result:

Add an example to show how to implement a json codec.

### CODE REVIEW

File: example/src/main/java/io/netty/example/gsonecho/GsonEchoServer.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import io.netty.util.CharsetUtil;
+import java.util.Map;
+
+/**
+ * Modification of {@link EchoServer} which utilizes Gson serialization.
+ */
+
+public class GsonEchoServer {
```
Comment (netkins) [inline example/src/main/java/io/netty/example/gsonecho/GsonEchoServer.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: example/src/main/java/io/netty/example/gsonecho/GsonEchoServer.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    static final boolean SSL = System.getProperty("ssl") != null;
+    static final int PORT = Integer.parseInt(System.getProperty("port", "8007"));
+
+    public static void main(String[] args) throws Exception {
```
Comment (netkins) [inline example/src/main/java/io/netty/example/gsonecho/GsonEchoServer.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: example/src/main/java/io/netty/example/gsonecho/GsonEchoClient.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.util.CharsetUtil;
+import java.util.Map;
+
+
+/**
+ * Modification of {@link EchoClient} which utilizes Gson serialization.
+ */
+public class GsonEchoClient {
```
Comment (netkins) [inline example/src/main/java/io/netty/example/gsonecho/GsonEchoClient.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: example/src/main/java/io/netty/example/gsonecho/GsonEchoClient.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    static final boolean SSL = System.getProperty("ssl") != null;
+    static final String HOST = System.getProperty("host", "127.0.0.1");
+    static final int PORT = Integer.parseInt(System.getProperty("port", "8007"));
+
+    public static void main(String[] args) throws Exception {
```
Comment (netkins) [inline example/src/main/java/io/netty/example/gsonecho/GsonEchoClient.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: example/src/main/java/io/netty/example/gsonecho/GsonEchoClient.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import io.netty.util.CharsetUtil;
+import java.util.Map;
+
+
+/**
+ * Modification of {@link EchoClient} which utilizes Gson serialization.
+ */
+public class GsonEchoClient {
+    static final boolean SSL = System.getProperty("ssl") != null;
+    static final String HOST = System.getProperty("host", "127.0.0.1");
```
Comment (netkins) [inline example/src/main/java/io/netty/example/gsonecho/GsonEchoClient.java]: MAJOR Make this IP "127.0.0.1" address configurable. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1313)

File: example/src/main/java/io/netty/example/gsonecho/GsonEchoClientHandler.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ */
+public class GsonEchoClientHandler extends ChannelInboundHandlerAdapter {
+
+    private final Map<String, String> message;
+
+    /**
+     * Creates a client-side handler.
+     */
+    public GsonEchoClientHandler() {
+        message = new HashMap<String, String>();
```
Comment (normanmaurer) [inline example/src/main/java/io/netty/example/gsonecho/GsonEchoClientHandler.java]: I think we should store it as an immutable map.

**Comment by windie** [comment]
Updated

**Comment by windie** [comment]
@normanmaurer - PTAL

**Comment by normanmaurer** [comment]
@windie sorry for the late response ... I totally lost track. Do you think this example will be beneficially for users? I think its just another example which does not something different then already existing ones. If you agree I would like to NOT pull this in as its just another dependency. Let me know.

**Comment by windie** [comment]
@normanmaurer sorry for the delay. Yes, I agreed that this one is not necessary. Closing it.

**Pull request closed without merging**